### PR TITLE
Combine `error` taming, rich `assert`, and `console` virtualization

### LINKED ISSUE #440: Features needed from `Error` + `console` + `assert`
See https://github.com/Agoric/agoric-sdk/pull/1619

First, need to move `assert` package from agoric-sdk to SES-shim.

Combine error taming, console taming + features, and assert into a new a group that shares some hidden state. 

If we can break up the error taming into a minimal first error taming to be safe (which might be the current one) and once SES supports vetted shims, we might make this triple an enhanced error,console,assert shim that replaces the tamed error. That would let us cleanly move it into its own package, and to omit it from minimal SES configurations.

The agoric-sdk assert package currently logs the actual data from the `details` template literal to the console when the error is thrown. Rather, in a manner similar to how the error-repair uses a weakmap on the side to associate stack trace info with an error, assert could likewise do so with the data from the `details` template literal. It should delay rendering this diagnostic data until that error is sent back to the console.

Our enhanced console, when getting an error object, should render any stack trace it held on the side in that weakmap, emulating the current de-facto typical console behavior on untamed errors, while keeping the stack hidden to those that just have the error. The old Caja/SES error taming did this.

Likewise, when getting an error object, our enhanced console should render the full details info it had kept on the side, if any, which keeping that data hidden.

It would generally be good to ask the console to associate hidden info with an error, to be rendered to the console only when the error is rendered. This would require adding a new non-standard api, perhaps to our console object. With such a generally available api, the new assert package would not need a special arrangement.

This could be used to keep stack traces associated with passed errors. Errors are passed by copy between mutually suspicious machines. Therefore anything hidden cannot be sent. The answer (with echoes in Causeway) is that the marshal layer on the sending side logs the error that it is sending, together with an extra marshal-generated identifier that it also sends in the serialized error record. The marshal on the receiving side asks its own console to associate this identifier with the received error. Someone, or some tool, that is given access to both logs can stitch them together --- can find the sending-side stack trace log that it uses to diagnose the stackless error as received.

### PR DESCRIPTION
**Pull request by erights**
Fixes #440

### CODE REVIEW

File: package.json
Comment (kriskowal) [inline package.json:7]: I have a conflicting change coming that changes this to `"workspaces": ["packages/*"]`.

File: packages/console/.prettierrc.json
Comment (kriskowal) [inline packages/console/.prettierrc.json:4]: This is a change we can and should apply to the entire repository, perhaps piecemeal and then holistically.

**Comment by erights** [comment]
Hi @kriskowal this is ready for review. However, in getting here, this got rather large. If you'd like, we should discuss how to break it up into reviewable-sized chunks.

**Review by kriskowal — COMMENTED** [review]
This is a partial review. I am still digesting.

**Comment by kriskowal** [comment]
> This is a partial review. I am still digesting.

Capturing out-of-band feedback: I’m recommending that we use an API like

```js
} catch (error) {
  assert.note(error, details`number too small ${num}`);
  throw error;
}
```

(Which I believe would be more palatable than `logToConsole(...asLogRecord(...))`)

Recommending that we use the catch-and-rethrow in SES module loading as a proving ground for this, and as an example of usage in this change.

I retract my suggestion that we use the term `cause` (in favor of using `note`), since I believe there’s a separate need for that term for in-band error annotation, e.g., `new Error('Could not find spoon', networkErrorThatCausedFailureToFindSpoon)`, where code can catch and programmatically inspect the cause chain to make relevant recovery decisions (like whether to retry).

**Comment by kriskowal** [comment]
I’m intrigued by the idea that the assertion library would provide “progressively enhanced” behavior depending on whether the `console` in scope is tamed or untamed. On the one hand, emitting the annotations to the debug system is good. On the other hand, reporting them to the untamed console, regardless of whether they’re later recovered, is a nightmare developer experience.

I will look more closely at this change. I suspect that the untamed case would *need* a mechanism for correlating related log messages, perhaps using a sequential identifier.

File: packages/assert/src/main.js
Comment (erights) [reply packages/assert/src/main.js:1]: I tried that but I couldn't get it to work. I don't know why. Attn @michaelfig

File: packages/console/src/assert.js
Comment (erights) [reply packages/console/src/assert.js]: Gone.

File: package.json
Comment (erights) [reply package.json]: reconciled.

File: packages/console/.prettierrc.json
Comment (erights) [reply packages/console/.prettierrc.json:4]: How do we apply it to the whole monorepo, rather than littering copies into each package?

File: packages/console/package.json
Comment (erights) [reply packages/console/package.json:5]: Done

File: packages/assert/src/main.js
Comment (michaelfig) [reply packages/assert/src/main.js:1]: You cannot "export" types that are JSDoc.  You need to put them in their own file (`packages/assert/src/types.js`), then make a `packages/assert/exported.js` that just contains `import './src/types';`

Your callers then use:
`import '@ses/assert/exported'` and all the types will be in scope.  If you are only using the API and not sharing types, the caller doesn't need to explicitly import them.

File: packages/console/test/throwsAndLogs.js
Comment (erights) [reply packages/console/test/throwsAndLogs.js:1]: Done

File: packages/console/.prettierrc.json
Comment (kriskowal) [reply packages/console/.prettierrc.json:4]: I’m not entirely certain about Prettier, but for Eslint, it was just a matter of moving this file up two directories. Might work. Might not. Otherwise, a symbolic link up two directories in each package.

Comment (kriskowal) [reply packages/console/.prettierrc.json:4]: Or, eventually, a package that we can depend upon in all of our projects in any repository.

File: packages/console/src/assert.js
Comment (kriskowal) [reply packages/console/src/assert.js]: @michaelfig clarified (on the LavaMoat call) that this debugger statement is intentional.

Comment (erights) [reply packages/console/src/assert.js]: So I should put it back in?

File: packages/console/package.json
Comment (michaelfig) [reply packages/console/package.json:23]: No, please don't do that.  The source files are the source of truth right now.  Typescript's conversion to `.d.ts` is lossy, so we would not preserve the types as they are right now.

However, you can crib a `jsconfig.json` from `agoric-sdk/packages/notifier`, and look at the `package.json` to see the `lint` scripts that use `tsc` to verify the types against the sources.  Add `"*.js"` to `"files"` (so that `exported.js` is available if present) and it will Just Work for both Typescript and JS consumers.

File: packages/console/src/console.js
Comment (michaelfig) [reply packages/console/src/console.js:3]: No, that won't work.  See above.

File: packages/console/src/assert.js
Comment (michaelfig) [reply packages/console/src/assert.js]: Please!  With a comment explaining that under a debugger, this statement will cause execution to pause where the developer can inspect the context.

**Comment by erights** [comment]
Ready for review!

**Comment by erights** [comment]
Adding @warner and leaving @michaelfig as ***optional*** reviewers. Only review from @kriskowal is critical.

File: packages/console/src/assert.js
Comment (erights) [reply packages/console/src/assert.js]: Done. But the comment does also say that leaving `debugger` in production code is controversial. The argument against: Under a debugger, you can breakpoint on a line without a `debugger` statement. However, AFAIK you cannot un-breakpoint a `debugger` statement.

Resolving though.

File: packages/console/package.json
Comment (erights) [reply packages/console/package.json:23]: With @michaelfig 's "please don't do that" I'm resolving this conversation.

File: packages/console/test/throwsAndLogs.js
Comment (erights) [reply packages/console/test/throwsAndLogs.js]: Done

**Comment by erights** [comment]
> > This is a partial review. I am still digesting.
> 
> Capturing out-of-band feedback: I’m recommending that we use an API like
> 
> ```js
> } catch (error) {
>   assert.note(error, details`number too small ${num}`);
>   throw error;
> }
> ```
> 
> (Which I believe would be more palatable than `logToConsole(...asLogRecord(...))`)

Done.

> Recommending that we use the catch-and-rethrow in SES module loading as a proving ground for this, and as an example of usage in this change.

Still TODO. After this PR is merged but before integration?

> I retract my suggestion that we use the term `cause` (in favor of using `note`), since I believe there’s a separate need for that term for in-band error annotation, e.g., `new Error('Could not find spoon', networkErrorThatCausedFailureToFindSpoon)`, where code can catch and programmatically inspect the cause chain to make relevant recovery decisions (like whether to retry).

Done.

File: packages/assert/src/assert.js
Comment (kriskowal) [inline packages/assert/src/assert.js:375]: Looks like we named `assertTypeof` to avoid a keyword identifier collision. Perhaps, like `equal` and `fail`, we don’t need this for `note`.

File: packages/assert/src/main.js
Comment (kriskowal) [reply packages/assert/src/main.js:1]: Do you mean literally `exported` or that whatever name we choose must correspond to a module that is in the `package.json` `"exports"` property?

```json
{
  "exports": {
    "./types": "./types.js"
  }
}
```

Such that `@agoric/assert/types` is externally visible?

File: packages/assert/src/assert.js
Comment (kriskowal) [inline packages/assert/src/assert.js:137]: It looks like `errorInfoKind` could be folded into the return of `getLogArgs` to simplify the signature of `rememberErrorInfo`.

Since `assert.note` is driving, it might be good for the console method to have a related name like `console.attachNote`, `deferNote`, or even just `note`.

Design note: I gather that the log array callback allows the `console` to return control to the `assert` module to expand details.

Maybe this could be more direct, like a `console.note(err, callback())`, where the only behavior is that the console will call the assertion library back if it encounters the given error object. Then, in the callback, `assert` would use the `console` API directly. Perhaps that would produce less specific coupling between `console` and `assert`.

File: packages/console/src/console.js
Comment (kriskowal) [inline packages/console/src/console.js:170]: Perhaps a good spot for a forward-looking options bag here.

File: packages/console/test/throws-and-logs.js
Comment (kriskowal) [inline packages/console/test/throws-and-logs.js:1]: I’m skeptical about `cycleTolerantStringify` being part of the committed public API of the `assert` module. I would go so far as to duplicate it rather than create a dependency. Or, just avoid using it.

File: packages/ses/package.json
Comment (kriskowal) [inline packages/ses/package.json:40]: These should both be `"^0.1.0"` once this is published. We can use version `0.0.0+1-dev` as the `"version"` in both of these for now, and make both of these dependency predicates `^0.0.0`. We bump these in the release process.

File: packages/ses/src/evaluate.js
Comment (kriskowal) [inline packages/ses/src/evaluate.js:4]: Perhaps it would make sense to export `assert`, `details`, and `quote`, then leave it to the consumer to import them by convention like:

```js
import { assert, details as d, quote as q } from '@agoric/assert';
```

This might be a good way to balance brevity and put the explanation as close to the confused reviewer as possible.

File: packages/ses/src/scope-handler.js
Comment (kriskowal) [inline packages/ses/src/scope-handler.js:26]: These are turning out well.

File: packages/ses/src/lockdown-shim.js
Comment (kriskowal) [inline packages/ses/src/lockdown-shim.js:123]: When we have a more complete story for vetted shims, does expressly threading and taming with our specific causal console dependency become less desirable? I imagine people will want to bring their own causal console implementation to the table. Is the expectation that they would implement a shim that replaces the global console before lockdown, which in turn gets wrapped by our causal console?

Another question: is our causal console coupled closely enough to `ses-shim` that no other package is likely to depend upon it?

**Review by kriskowal — CHANGES_REQUESTED** [review]
I would like to at least discuss the proposed `console` API extension. Apart from that, all remaining feedback is for your discretion.

File: packages/ses/package.json
Comment (michaelfig) [reply packages/ses/package.json:40]: We can't do that for `@agoric/assert`, as it is based on the `agoric-sdk`'s package which is already at `0.0.9`.  We could maybe make it `0.0.9+1-dev`.

**Review by michaelfig — COMMENTED** [review]
I looked carefully at the types, and from that standpoint, LGTM!

File: packages/assert/src/main.js
Comment (erights) [reply packages/assert/src/main.js]: @michaelfig fixed this in #461. Resolving.

File: packages/console/.prettierrc.json
Comment (erights) [reply packages/console/.prettierrc.json:4]: This change should not happen in this PR. Files #465. Resolving.

File: packages/console/test/throws-and-logs.js
Comment (erights) [reply packages/console/test/throws-and-logs.js:12]: Sorry, this was not intended to be a visible feature. I was only using it for my own debugging of my tests. Gone. Resolving.

File: packages/console/test/throws-and-logs.js
Comment (erights) [reply packages/console/test/throws-and-logs.js:36]: Done

File: packages/assert/src/assert.js
Comment (erights) [reply packages/assert/src/assert.js:135]: I didn't follow this. Let's discuss.

File: packages/console/src/console.js
Comment (erights) [reply packages/console/src/console.js:177]: Done

File: packages/ses/package.json
Comment (erights) [reply packages/ses/package.json:40]: I don't understand the suggestion. We should discuss. Until then I made both `"^0.1.0-dev"`. Not resolving.

File: packages/ses/src/evaluate.js
Comment (erights) [reply packages/ses/src/evaluate.js:4]: Done. Something very nice about this coupled use of `d` and `q` is that, at least in the font I'm using, they are vertical mirror images of each other.

File: packages/ses/src/lockdown-shim.js
Comment (erights) [reply packages/ses/src/lockdown-shim.js:123]: Let's discuss.

File: packages/assert/src/assert.js
Comment (erights) [reply packages/assert/src/assert.js]: > It looks like `errorInfoKind` could be folded into the return of `getLogArgs` to simplify the signature of `rememberErrorInfo`.
>
> Since `assert.note` is driving, it might be good for the console method to have a related name like `console.attachNote`, `deferNote`, or even just `note`.

`rememberErrorInfo` is gone

> Design note: I gather that the log array callback allows the `console` to return control to the `assert` module to expand details.

Callbacks work differently. PTAL

> Maybe this could be more direct, like a `console.note(err, callback())`, where the only behavior is that the console will call the assertion library back if it encounters the given error object. Then, in the callback, `assert` would use the `console` API directly. Perhaps that would produce less specific coupling between `console` and `assert`.

I did parameterize `makeCausalConsole` with a `loggedErrorHandler`, a privileged export from the `assert` package, so that neither `console` nor `assert` have any "extra" API surface. However, I did it differently than we talked about. I left the causal console in control of the chaining of error reports. The causal console primarily calls back into the `loggedErrorHandler` to access to `assert` module's hidden bookkeeping.

File: packages/ses/package.json
Comment (erights) [reply packages/ses/package.json]: Still unsure whether anything else needs to be done here. PTAL. Not resolving. Thanks.

File: packages/assert/src/assert.js
Comment (erights) [reply packages/assert/src/assert.js]: Not resolving yet due to the PTAL.

**Comment by erights** [comment]
Ready for review again. For `assert.js` and `console.js`, I recommend reviewing the whole files rather than the diffs.

**Comment by erights** [comment]
>  For assert.js and console.js, I recommend reviewing the whole files rather than the diffs.

Nevermind. I squashed and both modules are new compared to master, leaving little choice but to compare entire files. My apologies; I should have squashed less aggressively.

File: packages/ses/test/tame-console.test.js
Comment (kriskowal) [inline packages/ses/test/tame-console.test.js:62]: Is this TODO left intentionally?

File: packages/ses/test/tame-console.test.js
Comment (kriskowal) [inline packages/ses/test/tame-console.test.js:78]: TODO

**Review by kriskowal — APPROVED** [review]
Thank you for working in my feedback on the API. This seems better.

I am still concerned about the special relationship between `assert` and `console`. The `console` gets constructed with view into a particular `assert` module instance, but I expect each compartment will have its own `assert` module instance and errors will be forwarded only to the root `console`.

How do we ensure that details attached in a deeply nested compartment surface to the outermost console?

**Comment by erights** [comment]
Back to draft until it is again ready for review. Please spend no more time reviewing the current state.

File: packages/ses/test/tame-console.test.js
Comment (erights) [reply packages/ses/test/tame-console.test.js]: Done.

File: packages/ses/test/tame-console.test.js
Comment (erights) [reply packages/ses/test/tame-console.test.js]: Done.

File: packages/ses/package.json
Comment (erights) [reply packages/ses/package.json]: This PR no longer introduces new packages, so this question no longer arises. Resolving.

**Comment by erights** [comment]
Ready for review again! Includes new design doc at https://github.com/Agoric/SES-shim/tree/console/packages/ses/src/error

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:68]: TODO

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:48]: TODO

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:70]: Let’s link Causeway.

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:71]: Should link eventual-send.

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:1]: This is thorough!

Perhaps we can front-load some user-facing take-aways:

* Writing defensive programs under SES requires carefully considering what an error reveals to code positioned to catch those errors up the call chain.
* To that end, SES introduces an `assert` global with functions that that effect errors with annotations that will be hidden from callers, and tames the `Error` constructor such that it hides the `stack` to parent callers.
* SES tames the global `console` and grants it the ability to reveal error annotations and stacks to the actual console.
* Both `assert` and `console` are both powerful globals that SES does not implicitly carry into child compartments. When creating a child compartment, add `assert` to the compartment’s globals. Either add `console` too, or add a wrapper that annotates the console with a topic.
* SES hides annotations and stack traces by default. To reveal them, use a mechanism like `process.on("uncaughtException")` in Node.js to catch the error and log it back to the `console` tamed by `lockdown`.
* In the scope of the Agoric software ecosystem, this architecture will allow us to eventually introduce a more powerful distributed `console` that can meaningfully capture stack traces for a distributed debugger, based on the design of Causeway (linked).

We should have user-facing documentation for the usage of `assert` and `console`, including that last bit for revealing messages, either high in this README or in the SES README, until we have a better story for discovering SES documentation topics. My preference is the SES README so folks can use simple <kbd>⌘</kbd>+<kbd>F</kbd> to find relevant API docs.

File: packages/ses/src/error/README.md
Comment (kriskowal) [inline packages/ses/src/error/README.md:52]: TODO

File: packages/ses/src/evaluate.js
Comment (kriskowal) [inline packages/ses/src/evaluate.js:66]: TODO

File: packages/ses/src/module-load.js
Comment (kriskowal) [inline packages/ses/src/module-load.js:95]: TODO: Ticket for adding an argument to `fail` with the constructor for the error type.

File: packages/ses/test/error/assert-log.test.js
Comment (kriskowal) [inline packages/ses/test/error/assert-log.test.js:1]: TODO

File: packages/ses/test/error/assert.test.js
Comment (kriskowal) [inline packages/ses/test/error/assert.test.js:4]: TODO

**Review by kriskowal — APPROVED** [review]
Approved with nits. Thanks!

File: packages/ses/src/commons.js
Comment (erights) [reply packages/ses/src/commons.js:32]: Yup. According to https://ponyfoo.com/articles/polyfills-or-ponyfills it is a ponyfill. Revised comment accordingly.

File: packages/ses/src/commons.js
Comment (erights) [reply packages/ses/src/commons.js:22]: Done. Also cleaned up some other cases where we were destructuring `Object` to import from `commons` instead.

File: packages/ses/src/error/README.md
Comment (erights) [reply packages/ses/src/error/README.md:35]: I'm allergic to non-ascii characters appearing unescaped in source without compelling reason. I used `&mdash;` instead.

File: packages/ses/src/error/README.md
Comment (erights) [reply packages/ses/src/error/README.md:70]: Done

File: packages/ses/src/error/README.md
Comment (erights) [reply packages/ses/src/error/README.md:1]: Good text. I took it directly with only minor edits.

The "Logging Errors" section near the end of the SES README is already findable by searching on `assert` or `console` and already links to `src/error/README.md`, so I put this high level summary there. However, we don't yet have user-level documentation. I extracted that to https://github.com/Agoric/SES-shim/issues/486 and going to resolve this in place.

File: packages/ses/src/error/README.md
Comment (erights) [reply packages/ses/src/error/README.md:56]: I'm not sure whether we misunderstand each other. By TODO here I didn't mean that the documentation is incomplete. I mean that it is explaining future plans. I'll replace TODO language to clarify, but I think it is important to explain these future plans in this document. I will create a tracking issue to actually implement each major element of those plans.

Comment (erights) [reply packages/ses/src/error/README.md:56]: Done

File: packages/ses/src/error/README.md
Comment (erights) [reply packages/ses/src/error/README.md:52]: Gone. Subsumed by @warner 's https://github.com/Agoric/agoric-sdk/issues/1852 .

File: packages/ses/src/module-load.js
Comment (erights) [reply packages/ses/src/module-load.js:95]: Even better. Since that TODO was written I did that. But I hadn't noticed the opportunity to fix this line to use it. Done.

File: packages/ses/test/error/assert-log.test.js
Comment (erights) [reply packages/ses/test/error/assert-log.test.js:1]: Not sure what you want me to do here. In moving this package from the agoric-sdk monorepo to the SES-shim monorepo, I had to regress the testing to use tap or tape rather than AVA. I put this comment there to help when we reverse this regression, once SES-shim switches to use AVA.

I'll leave this unresolved, but I am not inclined to postpone merging till we resolve this.

File: packages/ses/test/error/assert.test.js
Comment (erights) [reply packages/ses/test/error/assert.test.js:4]: Likewise

File: packages/ses/src/evaluate.js
Comment (erights) [reply packages/ses/src/evaluate.js:66]: Left that line in but added a link to https://github.com/Agoric/SES-shim/issues/490
I think that resolves this well enough for now.

**Pull request merged**
